### PR TITLE
#5512: Log compile time templates/arguments

### LINKED ISSUE #5512: Log compile time templates/arguments
Have compile time templates/arguments printed in the build directory if compiling metal in debug mode. This helps while debugging instead of placing multiple prints to find the compile time arguments passed to the LLK functions. Currently this would affect performance, but it could be added to tests only when they are compiled in a "debug" mode. (Part of "Top Level LLK Debugging List" #4890)

### PR DESCRIPTION
**Pull request by ihamer-tt**
- Add logging to kernel_args.csv. Example:
```
transformer_attn_matmul/12827369245434441152/,ARCH_WORMHOLE,TENSIX_FIRMWARE,LOCAL_MEM_EN=0,UCK_CHLKC_UNPACK,NAMESPACE=chlkc_unpack,COMPILE_FOR_TRISC=0,KERNEL_BUILD,ARCH_WORMHOLE,TENSIX_FIRMWARE,LOCAL_MEM_EN=0,UCK_CHLKC_UNPACK,NAMESPACE=chlkc_unpack,COMPILE_FOR_TRISC=0,KERNEL_BUILD,ARG_0=0
writer_unary_interleaved_start_id/12956220480591854148/,ARCH_WORMHOLE,TENSIX_FIRMWARE,LOCAL_MEM_EN=0,COMPILE_FOR_BRISC,KERNEL_BUILD,ARCH_WORMHOLE,TENSIX_FIRMWARE,LOCAL_MEM_EN=0,COMPILE_FOR_BRISC,KERNEL_BUILD,NOC_INDEX=1,ARG_0=16,ARG_1=0
...
```
- Add python script to read constexprs from kernel ELF files. Example:
```
> python tt_metal/tools/dump-constexprs.py built/0/kernels/command_queue_producer/1129845549852061924/brisc/brisc.elf
Function: kernel_main
host_issue_queue_read_ptr_addr=0x0, issue_queue_start_addr=0x60, command_start_addr=0x1e000, data_section_addr=0x23800, data_buffer_size=0x14a800, consumer_cmd_base_addr=0x1e000, consumer_data_buffer_size=0xa2800
```

### CODE REVIEW

**Review by davorchap — APPROVED** [review]
This is cool! Perhaps constexprs dump in decimal?

**Comment by ihamer-tt** [comment]
I changed it to print both decimal and hex. Also, it now prints all constants as sometimes people do not use `constexpr`:

```
  host_issue_queue_read_ptr_addr = 0 (0x0)
  issue_queue_start_addr = 96 (0x60)
  issue_queue_size = 805306240 (0x2fffff80)
  command_start_addr = 122880 (0x1e000)
  data_section_addr = 145408 (0x23800)
  data_buffer_size = 1353728 (0x14a800)
  consumer_cmd_base_addr = 122880 (0x1e000)
  consumer_data_buffer_size = 665600 (0xa2800)
  consumer_noc_encoding = 48447231098880 (0x2c1000000000)
  pcie_core_noc_encoding = 13228499271680 (0xc0800000000)
  db_semaphore_addr = 109056 (0x1aa00)
```

File: tt_metal/jit_build/kernel_args.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+void dump_kernel_defines_and_args(const string &out_kernel_root_path) {
+    // Make sure the directory exists
+    tt::utils::create_file(out_kernel_root_path);
+
+    string kernel_args_csv = out_kernel_root_path + "kernel_args.csv";
+    ofstream file(kernel_args_csv, ios::trunc);
+    if (file.is_open()) {
+        std::lock_guard<std::mutex> lock(mutex_kernel_defines_and_args_);
```
Comment (pgkeller) [inline tt_metal/jit_build/kernel_args.cpp:48]: what's this lock guarding against?  should it be above the file open above?  if 2 threads can hit here, seems the 2nd could still clobber the first by truncating the file?

**Review by pgkeller — CHANGES_REQUESTED** [review]
I think we should allow this in release/production builds guarded behind an env var.  From 10K feet, I'd just use the watcher enable and call this a watcher feature.  I don't see anything here where the extra test for enabled/disabled would cause any perf concerns and that way noone needs to recompile to use it.

File: tt_metal/jit_build/kernel_args.cpp (hunk 1 above)
Comment (ihamer-tt) [reply tt_metal/jit_build/kernel_args.cpp:48]: It is against the kernel_defines_and_args_ map (it is shared across threads). You are right about the file, I will move the mutex to guard the file as well.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -212,7 +212,7 @@ If you are using a machine with bare metal machine specs, please use
   - Run `gdb --args python <python file> `
   - You can add breakpoints for future loaded libraries
 - To log the compile time arguments passed with `-D` during the kernel build phase:
-  - Build with debug symbols `make build CONFIG=debug`
+  - Run with Watcher enabled: `TT_METAL_WATCHER=120`
```
Comment (pgkeller) [inline CONTRIBUTING.md:215]: wonder if this should just be a link to the watcher section.  120 is a somewhat long update interval, but having it listed here means we have multiple places to change the docs to keep consistent.

**Review by pgkeller — APPROVED** [review]
one minor doc comment

**Review by tt-rkim — APPROVED** [review]

**Pull request merged**